libcaf_core/caf/event_based_mail.hpp

```diff
@@ -0,0 +1,142 @@
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/async_mail.hpp"
+#include "caf/event_based_response_handle.hpp"
+#include "caf/scheduled_actor.hpp"
+
+namespace caf {
+
+/// Provides a fluent interface for sending asynchronous messages to actors at a
+/// specific point in time.
+template <message_priority Priority, class Trait, class... Args>
+class event_based_scheduled_mail_t
+  : public async_scheduled_mail_t<Priority, Trait, Args...> {
+public:
+  using super = async_scheduled_mail_t<Priority, Trait, Args...>;
+
+  event_based_scheduled_mail_t(scheduled_actor* self, message&& content,
+                               actor_clock::time_point timeout)
+    : super(self, std::move(content), timeout) {
+    // nop
+  }
+
+  /// Sends the message to `receiver` as a request message and returns a handle
+  /// for processing the response.
+  /// @param receiver The actor that should receive the message.
+  /// @param ref_tag Either `strong_ref` or `weak_ref`. When passing
+  ///                `strong_ref`, the system will keep a strong reference to
+  ///                the receiver until the message has been delivered.
+  ///                Otherwise, the system will only keep a weak reference to
+  ///                the receiver and the message will be dropped if the
+  ///                receiver has been garbage collected in the meantime.
+  /// @param self_ref_tag Either `strong_self_ref` or `weak_self_ref`. When
+  ///                     passing `strong_self_ref`, the system will keep a
+  ///                     strong reference to the sender until the message has
+  ///                     been delivered. Otherwise, the system will only keep
+  ///                     a weak reference to the sender and the message will be
+  ///                     dropped if the sender has been garbage collected in
+  ///                     the meantime.
+  template <class Handle, class RefTag = strong_ref_t,
+            class SelfRefTag = strong_self_ref_t>
+  [[nodiscard]] auto request(const Handle& receiver, RefTag ref_tag = {},
+                             SelfRefTag self_ref_tag = {}) && {
+    detail::send_type_check<typename Trait::signatures, Handle, Args...>();
+    using response_type = response_type_t<typename Handle::signatures, Args...>;
+    auto mid = self()->new_request_id(Priority);
+    disposable in_flight;
+    if (receiver) {
+      auto& clock = self()->clock();
+      in_flight = clock.schedule_message(actor_cast(super::self_, self_ref_tag),
+                                         actor_cast(receiver, ref_tag),
+                                         super::timeout_, mid,
+                                         std::move(super::content_));
+
+    } else {
+      self()->enqueue(make_mailbox_element(self()->ctrl(), mid.response_id(),
+                                           make_error(sec::invalid_request)),
+                      self()->context());
```

Comment:
This should be simple enough to be covered by a unit test, delayed request to an invalid receiver.